refactor(web): improve code

hee-suh opened the pull request:
## 작업 내용

- ~로컬스토리지 값을 상수로 빼서 관리~
- 미들웨어 추가로 로그인한 사용자의 로그인 또는 회원가입 접근 제한
- addCommas 유틸 함수 리팩토링
    - for loop에서 reduce로 변경
- ~useEffect 대신 server component에서 api call [73e4826](https://github.com/ki99/popuniv/commit/73e482609af82f4c185e797ec5de85f1d4e33ba2)~
- useQuery hook 생성을 통해 반복되는 get 요청 -> state에 데이터 저장 작업 추상화
- http 유틸 함수형에서 클래스형으로 변경 시도!

hee-suh committed:
refactor(web): make localStorage values constant

hee-suh committed:
refactor(web): add middleware to restrict access
- prevent authenticated users from accessing the signin or signup page

hee-suh committed:
refactor(web): improve readability of addCommas
- change for loop to reduce
- rename addComma to addCommas

hee-suh committed:
refactor(web): use server component i/o useEffect
- get initial count from (home) server component instead of useEffect in the box.tsx

hee-suh committed:
refactor(web): implement useQuery hook

hee-suh committed:
wip(web): change http util from function to class

hee-suh commented:
@2-NOW

hee-suh commented on popuniv-web/app/signin/page.tsx:
```diff
@@ -11,10 +11,7 @@ import { SigninRequest, SigninResponse } from '../../models/interface';
 import { post } from '../../utils/http';
 
 const Signin = () => {
-	if (typeof window !== 'undefined' && localStorage.getItem('token')) {
```
로그인 직후, `router.push()`
로그인 이후 접근 시, middleware에서 `redirect`

hee-suh commented on popuniv-web/components/home/box.tsx:
```diff
@@ -1,72 +1,62 @@
 	const [count, setCount] = useState(0);
-	const [clickCount, setClickCount] = useState({ user: 0, all: 0 });
+	const [clickCount, setClickCount] = useState({ user: initialCount.userClickCount, all: initialCount.allClickCount });
+	// 로그인하지 않은 사용자의 Default University 클릭 횟수
```
클릭할 때마다 `accumulatedCount` 업데이트가 필요해서 상수화하지 않고 이대로 유지

hee-suh replied:
@2-NOW client component에서도 hydration은 발생해서요!

2-NOW replied:
아하.. useEffect를 쓰면 어떨까요 ?
부수효과를 리액트 라이프사이클에 동기화시킨다는 훅의 목적에도 맞아보여요

hee-suh replied:
useEffect를 쓰면 state도 별도로 만들어줘야 해서 가독성이 떨어진다고 생각했는데, typeof window !== undefined 보다는 useEffect를 사용하는 편이 더 낫나보네요...!

hee-suh replied:
순수하게 만들어주고, state로 관리를 해야겠네요..! `useEffect`를 사용하는 게 더 적절한 케이스에도 지양하자는 생각에 빠져버린 것 같네요.

추가로, `count`는 지금 당장 클릭하고 있는, 클릭할 때마다 하나씩 늘어나는 상태고, `clickCount`는 `count`가 끝나면 일괄 서버로 보낸 후 서버에서 받아온 값이고, `accumulatedCount`는 로그인하지 않은 사용자의 경우 로컬스토리지에 저장되는 값이라, 지금 상태에서는 나누는게 맞긴 한 거 같은데, `clickCount`를 useQuery에서 바로 처리하는 방안을 고려해봐야 할 것 같습니다.

감사합니다!!

hee-suh commented on popuniv-web/components/home/box.tsx:
```diff
@@ -1,72 +1,62 @@
 
 	const handleChangeGroup = (group: SelectOption) => {
-		if (!token) {
+		if (!TOKEN) {
```
새로고침 한 번 해야 TOKEN 인식되는 이슈

hee-suh commented on popuniv-web/hooks/common/useQuery.ts:
```diff
@@ -0,0 +1,36 @@
+		};
+
+		setLoading(true);
+		get<TResponse, TParam>({ url, param: JSON.parse(param) })
```
리렌더링 방지를 위해 문자열로 만들어줬던 param 다시 원상복구!

2-NOW replied:
여기도 param이 리렌더링이랑 어떤 연관인지 잘 모르겠어요

hee-suh commented on popuniv-web/middleware.ts:
```diff
@@ -0,0 +1,22 @@
+
+	const homeUrl = url.origin + '/';
+
+	if (token && isAuthUrl) {
```
로그인한 사용자 로그인 또는 회원가입 화면 접근 제한

2-NOW replied:
middleware에서 redirection 처리는 좋은 접근 같아요

hee-suh commented on popuniv-web/utils/http.ts:
```diff
@@ -1,6 +1,70 @@
 
 export const API_URL = process.env.NEXT_PUBLIC_API_URL;
+
+export class Http {
```
클래스로 만들어봤는데 좋은 use case가 따로 있는지?
get이랑 post도 더 리팩토링할 수 있을 것 같은데 묘안이 떠오르지 않음

2-NOW replied:
https://jasonwatmore.com/post/2020/04/18/fetch-a-lightweight-fetch-wrapper-to-simplify-http-requests
https://github.com/MaloLebrin/fetch-api-wrapper/blob/main/src/FetchWrapper.ts

hee-suh committed:
fix(web): change the token access location
from localStorage to cookies

hee-suh commented on popuniv-web/app/(home)/page.tsx:
```diff
@@ -1,12 +1,21 @@
 
 const Home = async () => {
-	// const [type, setType] = useState<GroupType>(Group.UNIVERSITY);
+	const group = USER && JSON.parse(USER)?.group;
```
token은 cookie에 저장되어 있어서 server actions로 가져왔는데, 
user는 쿠키에 저장할만한 값은 아닌 것 같고, 상수로 관리하면 업데이트가 되지 않아서 Home()에서 USER를 못 받아오는 것이 고민!

hee-suh replied:
Context API로 관리하겠습니다!

hee-suh commented on popuniv-web/components/home/list.tsx:
```diff
@@ -47,20 +47,11 @@ const Select = dynamic(() => import('react-select').then((mod) => mod.default),
-	}, []);
+	const { data } = useQuery<GroupInfo[], LeaderboardRequest>({
+		url: '/group',
+		param: JSON.stringify({ type: Group.UNIVERSITY }),
```
@2-NOW 의존성 배열에서 shallow equal로 비교하니까 object로 넘겨주게 되면 계속 다른 값으로 인식해버려서 primitive인 문자열로 변경했습니다!

2-NOW replied:
그럼 useQuery 외부에서 직렬화를 하기보단 훅 내부에서 직렬화를 하는것도 좋아보여요
외부에서 직렬화를 한다면 그 이유를 알기위해 내부구현을 까봐야만 하니까요

hee-suh replied:
엇 그렇게 하는게 훨씬 좋겠네요 감사합니다!

hee-suh committed:
revert(web): make localStorage values constant

hee-suh committed:
refactor(web): update useQuery & revert 73e4826
- use useQuery i/o server component and useEffect

hee-suh committed:
refactor(web): update useQuery param
- stringify param inside useQuery

hee-suh committed:
refactor(web): use type-only imports

hee-suh committed:
refactor(web): remove unnecessary async and await

hee-suh committed:
style(web): apply prettier
- change tabWidth from 4 to 2
- set useTabs false to convert tabs to spaces
- add importOrder settings

hee-suh committed:
style(web): change imports relative to absolute
- use absolute imports and module path aliases
